### PR TITLE
New solution to handle IVF_FLAT backward compatibility

### PR DESCRIPTION
**Pull request by cydrain**
Issue: #30

### CODE REVIEW

File: thirdparty/faiss/faiss/impl/index_read.cpp [hunk 1]
```diff
@@ -63,9 +63,17 @@ namespace faiss {
 static void read_index_header(Index* idx, IOReader* f) {
     READ1(idx->d);
     READ1(idx->ntotal);
+    READ1(idx->is_cosine);
+
+    uint8_t dummy8;
+    READ1(dummy8);
+    uint16_t dummy16;
```
Comment (foxspy) [inline thirdparty/faiss/faiss/impl/index_read.cpp:70]: 8(used for cosine) + 8 + 16 + 32 (rest padding) = 64 ? Pls add some comments here or just use 64 bit to save the `is_cosine`

File: thirdparty/faiss/faiss/impl/index_write.cpp [hunk 2]
```diff
@@ -81,9 +81,17 @@ namespace faiss {
 static void write_index_header(const Index* idx, IOWriter* f) {
     WRITE1(idx->d);
     WRITE1(idx->ntotal);
-    Index::idx_t dummy = 1 << 20;
-    WRITE1(dummy);
+    WRITE1(idx->is_cosine);
```
Comment (foxspy) [inline thirdparty/faiss/faiss/impl/index_write.cpp:84]: same as upper

File: thirdparty/faiss/faiss/impl/index_write.cpp [hunk 3]
```diff
@@ -397,6 +319,43 @@ void write_InvertedLists_with_norm(const InvertedLists* ils, IOWriter* f) {
...
+        WRITE1(h);
+        WRITE1(ils->nlist);
+        WRITE1(ils->code_size);
+        // this is a POD object
+        WRITEVECTOR(od->lists);
+
+        {
+            std::vector<OnDiskInvertedLists::Slot> v(
+                      od->slots.begin(), od->slots.end());
+            WRITEVECTOR(v);
```
Comment (foxspy) [inline thirdparty/faiss/faiss/impl/index_write.cpp:352]: Has IVFFLAT used these inverted lists? May beyond my knowledge

File: src/common/utils.cc [hunk 4]
```diff
@@ -83,28 +83,39 @@ ConvertIVFFlatIfNeeded(const BinarySet& binset, const uint8_t* raw_data, const s
...
+        ivfl->code_size = ivfl->d * sizeof(float);
+
+        auto remains = binary->size - reader.tellg() - sizeof(uint32_t) - sizeof(ivfl->invlists->nlist) -
+                       sizeof(ivfl->invlists->code_size);
+        auto invlist_size = sizeof(uint32_t) + sizeof(size_t) + ivfl->nlist * sizeof(size_t);
+        auto ids_size = ivfl->ntotal * sizeof(faiss::Index::idx_t);
+        // auto codes_size = ivfl->d * ivfl->ntotal * sizeof(float);
+
+        // IVF_FLAT_NM format, need convert to new format
+        if (remains == invlist_size + ids_size) {
```
Comment (foxspy) [inline src/common/utils.cc:102]: little hack to decide the index type by binary size. Is there any other ways?

**Comment by foxspy** [comment]
Pls verify the compatibility with existing index binary. A byte difference may occur load fail. @elstic

File: src/common/utils.cc (hunk 4 above)
Comment (cydrain) [reply src/common/utils.cc:102]: I only think out 2 ways to detect IVF_FLAT_NM:
1. use native IVF_FLAT format to parse a binary, if it throw an exception, then this binary should be IVF_FLAT_NM
2. the way as this PR, read ivf_flat index header, then calculate the binary size

before this PR, I use method 1, but it has a little risk, so I change to method 2

File: thirdparty/faiss/faiss/impl/index_read.cpp (hunk 1 above)
Comment (cydrain) [reply thirdparty/faiss/faiss/impl/index_read.cpp:70]: yes

**Pull request merged**